### PR TITLE
PPTP-1459 | API to surface financial data

### PR DESCRIPTION
**Pull request by liampan**
### Description of Work carried through

API for fincails used by PPT and BTA to render Tile info

#### Check list 
 - [ ] `./precheck` was executed (Integration/Component/Unit tests)
 - [ ] `sbt scalafmt test:scalafmt` was executed
 - [ ] Required Environment Config has been amended/added
 - [ ] Links to dependencies have been included (BE/FE work)
 - [ ] User Acceptance Tests (UAT) were run locally and have passed.

### CODE REVIEW

File: app/uk/gov/hmrc/plasticpackagingtaxreturns/services/PPTObligationsService.scala [hunk 1]
```diff
@@ -43,12 +42,12 @@ class PPTObligationsService {
   private def construct(obligation: Obligation): PPTObligations = {
     val today: LocalDate = LocalDate.now()
```
Comment (ajchilts) [inline app/uk/gov/hmrc/plasticpackagingtaxreturns/services/PPTObligationsService.scala:43]: Do we need to use UTC for server time difference?

File: app/uk/gov/hmrc/plasticpackagingtaxreturns/controllers/PPTFinancialsController.scala [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  financialsService: PPTFinancialsService
+)(implicit val executionContext: ExecutionContext)
+    extends BackendController(cc) {
+
+  def get(ref: String): Action[AnyContent] =
+    authenticator.authorisedAction(parse.default) {
+      implicit request =>
+        financialDataConnector.get(ref,
+                                   appConfig.pptTaxStartDate,
+                                   LocalDate.now(),
```
Comment (ajchilts) [inline app/uk/gov/hmrc/plasticpackagingtaxreturns/controllers/PPTFinancialsController.scala:46]: Do we need to use UTC for server time difference?

File: app/uk/gov/hmrc/plasticpackagingtaxreturns/controllers/PPTObligationsController.scala [hunk 3]
```diff
@@ -44,7 +45,7 @@ class PPTObligationsController @Inject() (
   def get(ref: String): Action[AnyContent] =
     authenticator.authorisedAction(parse.default) {
       implicit request =>
-        obligationsDataConnector.get(ref, PPTTaxStartDate, LocalDate.now(), ObligationStatus.OPEN).map {
+        obligationsDataConnector.get(ref, appConfig.pptTaxStartDate, LocalDate.now(), ObligationStatus.OPEN).map {
```
Comment (ajchilts) [inline app/uk/gov/hmrc/plasticpackagingtaxreturns/controllers/PPTObligationsController.scala:48]: Same for all references to LocalDate.now()

File: app/uk/gov/hmrc/plasticpackagingtaxreturns/config/AppConfig.scala [hunk 4]
```diff
@@ -19,12 +19,15 @@ package uk.gov.hmrc.plasticpackagingtaxreturns.config
...
 import uk.gov.hmrc.play.bootstrap.config.ServicesConfig
 
+import java.time.LocalDate
 import javax.inject.{Inject, Singleton}
 import scala.concurrent.duration.FiniteDuration
 
 @Singleton
 class AppConfig @Inject() (config: Configuration, servicesConfig: ServicesConfig) {
 
+  val pptTaxStartDate: LocalDate = LocalDate.of(2022, 4, 1)
```
Comment (kiranprakashdigital) [inline app/uk/gov/hmrc/plasticpackagingtaxreturns/config/AppConfig.scala:29]: Should we have this value in the conf file?

**Review by kiranprakashdigital — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/plasticpackagingtaxreturns/services/PPTFinancialsService.scala [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+class PPTFinancialsService {
+
+  def construct(data: FinancialDataResponse): PPTFinancials = {
+    val charges: Seq[Charge] = data.financialTransactions.map(Charge.apply)
+
+    val dueCharge: Option[Charge] =
+      charges.filter(_.due.isEqualOrAfterToday).sortBy(_.due).headOption
+
+    val overdueCharges: Seq[Charge] =
+      charges.filter(_.due.isBeforeToday).sortBy(_.due)
```
Comment (lmcewanhmrc) [inline app/uk/gov/hmrc/plasticpackagingtaxreturns/services/PPTFinancialsService.scala:37]: Don't need to sort?

File: app/uk/gov/hmrc/plasticpackagingtaxreturns/services/PPTFinancialsService.scala [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+    if (totalChargesSum < 0) PPTFinancials.inCredit(totalChargesSum)
+    else
+      (debitAmount, overdueAmount) match {
+        case (None, None)                         => PPTFinancials.NothingDue
+        case (Some((amount, due)), None)          => PPTFinancials.debitDue(amount, due)
+        case (None, Some(amount))                 => PPTFinancials.overdue(amount)
+        case (Some((amount, due)), Some(overdue)) => PPTFinancials.debitAndOverdue(amount, due, overdue)
+      }
+  }
```
Comment (lmcewanhmrc) [inline app/uk/gov/hmrc/plasticpackagingtaxreturns/services/PPTFinancialsService.scala:56]: I think this method ^^^ can be simplified - let's get together and discuss

File: test/uk/gov/hmrc/plasticpackagingtaxreturns/services/PPTFinancialsServiceSpec.scala [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+  def makeData(charges: (BigDecimal, LocalDate)*): FinancialDataResponse =
+    emptyData.copy(financialTransactions =
+      charges.map {
+        case (amount, due) =>
+          emptyTransaction.copy(outstandingAmount = Some(amount), items = Seq(emptyItem.copy(dueDate = Some(due))))
+      }
+    )
+
+  "construct" must {
+    "return NothingDue" when {
```
Comment (lmcewanhmrc) [inline test/uk/gov/hmrc/plasticpackagingtaxreturns/services/PPTFinancialsServiceSpec.scala:51]: This is test is when there are no financials returned? Nothing due is another test where sum of amountOutstanding == 0?

File: test/uk/gov/hmrc/plasticpackagingtaxreturns/services/PPTFinancialsServiceSpec.scala [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+      "there is no FinancialTransactions" in {
+        sut.construct(makeData()) mustBe PPTFinancials.NothingDue
```
Comment (lmcewanhmrc) [inline test/uk/gov/hmrc/plasticpackagingtaxreturns/services/PPTFinancialsServiceSpec.scala:53]: Is NothingDue (None, None, None)? (Just x-ref-ing with wiki page...)

File: test/uk/gov/hmrc/plasticpackagingtaxreturns/services/PPTFinancialsServiceSpec.scala [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+      "overdue in credit" in {
+        sut.construct(makeData(credit -> yesterday)) mustBe PPTFinancials.inCredit(amount)
+      }
+      "overdue in credit more than due is due" in {
+        val fullCredit = BigDecimal.valueOf(Double.MinValue)
+        sut.construct(makeData(fullCredit -> yesterday, amount -> today)) mustBe PPTFinancials.inCredit(
+          fullCredit + amount
+        )
+      }
+    }
```
Comment (lmcewanhmrc) [inline test/uk/gov/hmrc/plasticpackagingtaxreturns/services/PPTFinancialsServiceSpec.scala:103]: There's possibly more test cases that should be here... but let's discuss...

**Review by lmcewanhmrc — CHANGES_REQUESTED** [review]

File: test/uk/gov/hmrc/plasticpackagingtaxreturns/services/PPTFinancialsServiceSpec.scala (hunk 8 above)
Comment (liampan) [reply test/uk/gov/hmrc/plasticpackagingtaxreturns/services/PPTFinancialsServiceSpec.scala:53]: Yes, open to better name suggestions

Comment (liampan) [reply test/uk/gov/hmrc/plasticpackagingtaxreturns/services/PPTFinancialsServiceSpec.scala:53]: `NothingOutstanding` I guess is more correct?

File: app/uk/gov/hmrc/plasticpackagingtaxreturns/config/AppConfig.scala (hunk 4 above)
Comment (liampan) [reply app/uk/gov/hmrc/plasticpackagingtaxreturns/config/AppConfig.scala:29]: Yes to avoid duplication

Comment (kiranprakashdigital) [reply app/uk/gov/hmrc/plasticpackagingtaxreturns/config/AppConfig.scala:29]: Can you move it to conf file, we have something similar in registration-frontend

**Review by lmcewanhmrc — APPROVED** [review]

File: app/uk/gov/hmrc/plasticpackagingtaxreturns/config/AppConfig.scala (hunk 4 above)
Comment (liampan) [reply app/uk/gov/hmrc/plasticpackagingtaxreturns/config/AppConfig.scala:29]: But its not configurable per env? is that not just adding it for the sake of adding it?

**Pull request merged**